concert/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+const NFT_COUNT: u128 = 1;
+
+#[derive(Debug, Default)]
+pub struct Concert {
+    pub owner_id: ActorId,
+    pub contract_id: ActorId,
+
+    pub name: TokenId,
+    pub creator: ActorId,
+    pub number_of_tickets: u128,
+    pub date: u128,
+
+    pub buyers: BTreeSet<ActorId>,
+
+    pub id_counter: u128,
+    // user to token id to metadata
+    pub metadata: BTreeMap<ActorId, BTreeMap<u128, Option<TokenMetadata>>>,
+}
+
+static mut CONTRACT: Option<Concert> = None;
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let config: InitConcert = msg::load().expect("Unable to decode InitConfig");
+    let concert = Concert {
+        owner_id: config.owner_id,
+        contract_id: config.mtk_contract,
+        ..Default::default()
+    };
+    CONTRACT = Some(concert);
+}
+
+#[gstd::async_main]
+async unsafe fn main() {
+    let action: ConcertAction = msg::load().expect("Could not load Action");
+    let concert: &mut Concert = unsafe { CONTRACT.get_or_insert(Concert::default()) };
+    match action {
+        ConcertAction::Create {
+            creator,
+            concert_id,
+            no_tickets,
+            date,
+        } => concert.create_concert(creator, concert_id, no_tickets, date),
+        ConcertAction::Hold { concert_id } => concert.hold_concert(concert_id).await,
+        ConcertAction::BuyTickets {
+            concert_id,
+            amount,
+            metadata,
+        } => concert.buy_tickets(concert_id, amount, metadata).await,
+    }
+}
+
+impl Concert {
+    fn create_concert(
+        &mut self,
+        creator: ActorId,
+        concert_id: u128,
+        number_of_tickets: u128,
+        date: u128,
+    ) {
+        self.creator = creator;
+        self.id_counter = concert_id;
+        self.name = concert_id;
+        self.number_of_tickets = number_of_tickets;
+        self.date = date;
+        msg::reply(
+            ConcertEvent::Creation {
+                creator,
+                concert_id,
+                no_tickets: number_of_tickets,
+                date,
+            },
+            0,
+        )
+        .unwrap();
+    }
+
+    async fn buy_tickets(
+        &mut self,
+        concert_id: u128,
+        amount: u128,
+        mtd: Vec<Option<TokenMetadata>>,
+    ) {
+        if msg::source() == ZERO_ID {
+            panic!("CONCERT: Message from zero address");
+        }
+
+        if amount < 1 {
+            panic!("CONCERT: Can not buy less than 1 ticket");
+        }
+
+        if self.number_of_tickets < amount {
+            panic!("CONCERT: Not enought tickets");
+        }
+
+        if mtd.len() != amount as usize {
+            panic!("CONCERT: Metadata not provided for all the tickets");
+        }
+
+        for meta in mtd.iter() {
+            self.id_counter += 1;
+            self.metadata
+                .entry(msg::source())
+                .or_default()
+                .insert(self.id_counter + 1, (*meta).clone());
```

Comment:
No need to clone if we consume `mtd`.
```suggestion
                .insert(self.id_counter + 1, meta);
```